### PR TITLE
Control by disableUnsafeDebugFlag

### PR DESCRIPTION
**Pull request by kjeom**
## Proposed changes

- This is a follow-up PR to #1672
- This PR introduces the `rpc.unsafe-debug.disable` flag
- If you set the `rpc.unsafe-debug.disable` flag, 
  - can use only a subset of debug APIs. Refer to `Further Comments` for the full list
  - cannot use the [javascript-based tracing](https://ko.docs.klaytn.foundation/dapp/json-rpc/api-references/debug/tracing#javascript-based-tracing) for [VM tracing](https://docs.klaytn.foundation/content/dapp/json-rpc/api-references/debug/tracing) APIs, but only [pre-defined tracers](https://ko.docs.klaytn.foundation/dapp/json-rpc/api-references/debug/tracing#tracing-options) are allowed (e.g., `revertTracer`)
- How to set the flag
  - add ADDITIONAL="--rpc.unsafe-debug.disable" in kxxd.conf file

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### Why this new flag `rpc.unsafe-debug.disable` is introduced and why some debug APIs are disabled by that flag?
- The debug APIs are for private use in nature, however some public ENs provide debug namespace over RPC for user convenience
- Some debug APIs are vulnerable. Some debug APIs are too heavy/costly. Some debug APIs are meaningless to be provided over RPC.
- Nevertheless, we don't want to discourage public ENs to provide some useful & relatively safe debug APIs (e.g., `debug_traceTransaction`)

### Full list of APIs affected
- debug APIs that are disabled when `rpc.unsafe-debug.disable` flag was set
  - debug_printBlock
  - debug_setHead
  - debug_startWarmUp, debug_startContractWarmUp, debug_stopWarmUp
  - debug_startCollectingTrieStats
  - debug_chaindbProperty, debug_chaindbCompact
  - debug_verbosity, debug_verbosityByName, debug_verbosityByID
  - debug_vmodule, debug_backtraceAt, debug_stacks, debug_freeOSMemory, debug_setGCPercent, debug_memStats, debug_gcStats, debug_startPProf, debug_stopPProf, debug_isPProfRunning, debug_cpuProfile, debug_startCPUProfile, debug_stopCPUProfile
  - debug_goTrace, debug_startGoTrace, debug_stopGoTrace
  - debug_blockProfile, debug_setBlockProfileRate, debug_writeBlockProfile, debug_mutexProfile, debug_setMutexProfileRate, debug_writeMutexProfile, debug_writeMemProfile
  - debug_standardTraceBadBlockToFile, debug_standardTraceBlockToFile
  - debug_preimage
  - debug_storageRangeAt
  - debug_setVMLogTarget

- debug APIs that are NOT disabled even if `rpc.unsafe-debug.disable` flag was set
  - [VM tracing](https://docs.klaytn.foundation/content/dapp/json-rpc/api-references/debug/tracing) APIs, however with limited functionality (only [pre-defined tracers](https://ko.docs.klaytn.foundation/dapp/json-rpc/api-references/debug/tracing#tracing-options) are allowed)
    - debug_traceBadBlock, debug_traceBlock, debug_traceBlockByHash, debug_traceBlockByNumber, debug_traceBlockByNumberRange, debug_traceBlockFromFile, debug_traceTransaction, debug_traceChain
  - debug_dumpBlock, debug_dumpStateTrie, debug_getBlockRlp, debug_getModifiedAccountsByHash, debug_getModifiedAccountsByNumber, debug_getBadBlocks, debug_getModifiedStorageNodesByNumber
  - debug_metrics

### CODE REVIEW

**Comment by kjeom** [comment]
@2dvorak I think we need to test https://github.com/klaytn/klaytn/issues/1676#issuecomment-1316106070 again.

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -71,6 +71,11 @@ var (
 	ConfFlag = cli.StringFlag{
 		Name: "conf",
 	}
+	UnsafeDebugDisableFlag = cli.BoolFlag{
+		Name:   "unsafe-debug.disable",
```
Comment (aidan-kwon) [inline cmd/utils/flags.go:75]: There are other flags starting with "rpc". How about changing its name to "rpc.debug.disable-critical" or "rpc.debug.disable-unsafe"?

File: cmd/utils/nodecmd/nodeflags.go [hunk 2]
```diff
@@ -283,6 +283,7 @@ var KENFlags = []cli.Flag{
 	altsrc.NewUint64Flag(utils.TxResendIntervalFlag),
 	altsrc.NewIntFlag(utils.TxResendCountFlag),
 	altsrc.NewBoolFlag(utils.TxResendUseLegacyFlag),
+	altsrc.NewBoolFlag(utils.UnsafeDebugDisableFlag),
```
Comment (aidan-kwon) [inline cmd/utils/nodecmd/nodeflags.go:286]: I think the flag is supposed to be `CommonRPCFlags"

File: node/node.go [hunk 3]
```diff
@@ -802,6 +794,21 @@ func (n *Node) apis() []rpc.API {
 			Public:    true,
 		},
 	}
+	debugRpcApi := []rpc.API{
+		{
+			Namespace: "unsafedebug",
```
Comment (aidan-kwon) [inline node/node.go:799]: Not "debug"?

File: node/node.go [hunk 4]
```diff
@@ -802,6 +794,21 @@ func (n *Node) apis() []rpc.API {
...
+			Version:   "1.0",
+			Service:   debug.Handler,
+		}, {
+			Namespace: "unsafedebug",
```
Comment (aidan-kwon) [inline node/node.go:803]: Not "debug"?

File: cmd/utils/flags.go (hunk 1 above)
Comment (kjeom) [reply cmd/utils/flags.go:75]: Make sense. but I was thinking about the disable convention xxx.disable for hierarchy after applied yaml configuration. ref. [example.yaml](https://github.com/klaytn/klaytn/pull/1684/files#diff-6c1ff5ac872c0749f63ab487947ccc6140b5b9ee5c1b692de5c1511121b4d44cR15).

Then how about `rpc.unsafe-debug.disable`?

File: cmd/utils/nodecmd/nodeflags.go (hunk 2 above)
Comment (kjeom) [reply cmd/utils/nodecmd/nodeflags.go]: Thanks, resolved [b83e083](https://github.com/klaytn/klaytn/pull/1746/commits/b83e0832b21a60dfb66c7f1b9793825311825111)

File: node/node.go (hunk 3 above)
Comment (kjeom) [reply node/node.go]: Right, renamed [f34a3f3](https://github.com/klaytn/klaytn/pull/1746/commits/f34a3f3117c0119cb1538ca0c26b3e1fa18dc9d2)

File: node/node.go (hunk 4 above)
Comment (kjeom) [reply node/node.go]: Right, renamed [f34a3f3](https://github.com/klaytn/klaytn/pull/1746/commits/f34a3f3117c0119cb1538ca0c26b3e1fa18dc9d2)

**Comment by blukat29** [comment]
Please fix `cmd/utils/nodecmd/consolecmd_test.go:ipcAPIs` as well.
To check the fix, you'll have to manually run test (because `nodecmd` is ignored in `make test-others`).

```
env GOPATH=$(go env GOPATH) GO111MODULE=on go run build/ci.go test -p 1 ./cmd/utils/nodecmd/
```

File: cmd/utils/flags.go (hunk 1 above)
Comment (aidan-kwon) [reply cmd/utils/flags.go:75]: Looks good to me

**Review by JayChoi1736 — DISMISSED** [review]

File: node/cn/tracers/api.go [hunk 5]
```diff
@@ -109,23 +109,23 @@ func NewUnsafeAPI(backend Backend) *API {
 }
 
 type chainContext struct {
-	api *API
```
Comment (2dvorak) [inline node/cn/tracers/api.go:112]: Is this change necessary, or a minor improvement?

**Comment by 2dvorak** [comment]
Although the `--rpc.unsafe-debug.disable` flag was set and unsafe-debug APIs are blocked, the node console still shows all debug API list. Can we fix that too?

```
> debug
{
  backtraceAt: function(),
  blockProfile: function(),
  chaindbCompact: function(),

... more list ...

  writeMemProfile: function(),
  writeMutexProfile: function()
}
> debug.backtraceAt("")
Error: the method debug_backtraceAt does not exist/is not available
    at web3.js:3278:20
    at web3.js:6810:15
    at web3.js:5221:36
    at <anonymous>:1:1
```

File: node/node.go [hunk 6]
```diff
@@ -802,6 +794,21 @@ func (n *Node) apis() []rpc.API {
...
 	}
+	debugRpcApi := []rpc.API{
+		{
+			Namespace: "debug",
+			Version:   "1.0",
+			Service:   debug.Handler,
+		}, {
+			Namespace: "debug",
+			Version:   "1.0",
+			Service:   NewPublicDebugAPI(n),
```
Comment (aidan-kwon) [inline node/node.go:805]:
```suggestion
			Service:   NewPublicDebugAPI(n),
			Public: true,
```

File: node/cn/backend.go [hunk 7]
```diff
@@ -487,13 +487,26 @@ func (s *CN) APIs() []rpc.API {
...
+		tracerAPI = tracers.NewUnsafeAPI(s.APIBackend)
+		apis = append(apis, []rpc.API{
+			{
+				Namespace: "debug",
+				Version:   "1.0",
+				Service:   NewPrivateDebugAPI(s.chainConfig, s),
+				Public:    false,
+			},
+		}...)
+	}
```
Comment (aidan-kwon) [inline node/cn/backend.go:508]: It would be more clear if you change method names like this:
 - `NewAPI` -> `NewAPIUnsafeDisabled`
 - `NewUnsafeAPI` -> `NewAPI`

File: node/node.go (hunk 6 above)
Comment (aidan-kwon) [reply node/node.go:805]: Refer to https://github.com/klaytn/klaytn/blob/8664bc88319567f560bc5244131672d554e55293/node/node.go#L786-L790

Comment (2dvorak) [reply node/node.go:805]: The terms are misleading, but..

- A `Public` service would be enabled if there is no namespace specified in the `--rpcapi` flag.
- The `PublicDebugAPI` refers to "not unsafe" debug APIs that no vulnerabilities are found.. **yet**.

So I removed the `Public` property because I thought it would be better to disable debug APIs if not explicitly specified.

File: node/cn/tracers/api.go (hunk 5 above)
Comment (kjeom) [reply node/cn/tracers/api.go:112]: Yes, It's a minor improvement for calling the `newChainContext` private function with no API instance (not receiver function).

**Review by aidan-kwon — APPROVED** [review]
LGTM except for minors

File: node/cn/backend.go (hunk 7 above)
Comment (kjeom) [reply node/cn/backend.go]: Resolved [8026919](https://github.com/klaytn/klaytn/pull/1746/commits/80269194f01f7e8fff13cf1ee335e4ca0876d5da)

File: node/cn/tracers/api.go (hunk 5 above)
Comment (2dvorak) [reply node/cn/tracers/api.go:112]: OK, thank you.

File: node/cn/tracers/api.go [hunk 8]
```diff
@@ -98,13 +98,13 @@ type API struct {
 
 // NewAPI creates a new API definition for the tracing methods of the CN service,
```
Comment (aidan-kwon) [inline node/cn/tracers/api.go:99]: In the comment, `NewAPI` -> `NewAPIUnsafeDisabled`

File: node/cn/tracers/api.go [hunk 9]
```diff
@@ -98,13 +98,13 @@ type API struct {
...
 // only allowing predefined tracers.
-func NewAPI(backend Backend) *API {
+func NewAPIUnsafeDisabled(backend Backend) *API {
 	return &API{backend: backend, unsafeTrace: false}
 }
 
 // NewUnsafeAPI creates a new API definition for the tracing methods of the CN service,
```
Comment (aidan-kwon) [inline node/cn/tracers/api.go:105]: In the comment, `NewUnsafeAPI` -> `NewAPI`

File: node/cn/tracers/api.go (hunk 8 above)
Comment (kjeom) [reply node/cn/tracers/api.go]: Resolved cf88fea

File: node/cn/tracers/api.go (hunk 9 above)
Comment (kjeom) [reply node/cn/tracers/api.go]: Resolved cf88fea

**Review by aidan-kwon — APPROVED** [review]

**Review by 2dvorak — APPROVED** [review]
